Cython/Compiler/Nodes.py

```diff
@@ -7707,7 +7707,8 @@ def generate_execution_code(self, code):
         code.mark_pos(self.pos)
         old_loop_labels = code.new_loop_labels()
         self.iterator.generate_evaluation_code(code)
-        code.putln("for (;;) {")
+        if not self.iterator.has_custom_for_loop:
```

Comment:
I wonder if it'd be better to do

````
code.put("for (")
self.iterator.generate_for_loop_header(code)
code.put(") {")
```

with the default version being

```
def generate_for_loop_header(self, code):
    code.put(";;")
```
?

What I don't really like about your current version is that it's a little unclear where the `for (...) {` is generated in the `has_custom_for_loop` case.

Reply:
Totally agree, but I'm not sure where to put `generate_for_loop_header` default implementation. 
The common ancestor of IteratorNode, CppIteratorNode, AsyncIteratorNode is ExprNode. Is it ok if we put it there?

Reply:
`ExprNode` is too general I think.

Given how simple the default implementation is, you could just duplicate it on `IteratorNode` and `AsyncIteratorNode` and not worry about inheritance.